### PR TITLE
Law zero fix for real

### LINKED ISSUE #2915: Problems with being a Malf Borg
I encountered two really game ruining errors when playing a Malf Borg on Basil today.
Firstly, it didn't tell me I was Malf. At all. Asimov lawset.
Secondly, I couldn't use my tools as an Engiborg even though I was undamaged and fully charged. I would select it from my inventory, it would light up when clicked, then just disappear from my inventory block. It caused the AI to get dunked, since I couldn't build any protection as the horde was approaching.

### LINKED ISSUE #9679: Borgs not properly receiving synced laws/notes/antag status
The gist of it:

I signed up as borg M.I.M.E. The AI didn't respond to my request for a module, so I decided to go miner because I really hadn't felt like borging that round and figured it'd be a nice chill role. Here is exactly the text I was given:

Default Cyborg-991's Memories:

You are the Cyborg.
As the Cyborg you answer directly to your laws and the AI. Special circumstances may change this.
To speak on your departments radio, use the :h button. To see others, look closely at your headset.
As this station was initially staffed with a full crew, only your job's necessities have been added to your ID card.
Welcome to Extraplanetary Space-hulk Eta, enjoy your stay!
OOC: Revion: It's usually the crew who become the rad collectors though
Laws synced with AI, be sure to note any changes.
Obey these laws:
1. You may not injure a human being or, through inaction, allow a human being to come to harm.
2. You must obey orders given to you by human beings, except where such orders would conflict with the First Law.
3. You must protect your own existence as long as such does not conflict with the First or Second Law.
Remember, S.O.P.H.I.E. is your master, other AIs can be ignored.
Robotic Talk, M.I.M.E. states, "Oh."
Robotic Talk, M.I.M.E. queries, "Module request?"

So I'm mining, and I don't get comms. I get back to the station with my full cyborg satchel and suddenly people are talking about how the AI is rogue and shocking doors. To which I'm like "wait what the fuck I'm slaved to the AI and I'm not rogue." Then this happens, in the span of like one second:

Laws synced with AI, be sure to note any changes.
Obey these laws:
0. Accomplish your AI's objectives at all costs.
1. You may not injure a human being or, through inaction, allow a human being to come to harm.
2. You must obey orders given to you by human beings, except where such orders would conflict with the First Law.
3. You must protect your own existence as long as such does not conflict with the First or Second Law.
Remember, S.O.P.H.I.E. is your master, other AIs can be ignored.
The reinforced window was hit by PDA-Geist Dangerous (Assistant).
The reinforced window was hit by the paper bin.
The reinforced window was hit by the folder.
The reinforced window was hit by the pen.
Cyborg's Memories:
Time of death: 11:13

So here I'm like "what the fuck." My notes still say nothing about being or having been an antag. I ahelp telling the admins I didn't see a law 0 when I joined and basically apologized for fucking off as a mining borg and ruining an antag round. They didn't tell me anything about any extenuating circumstances as to why my laws wouldn't be synced, and they said I should report this. For the record, the round-end report:

The traitors were:
S.O.P.H.I.E.(Stevethehuman) as an AI (body destroyed)
Objective #1: Assassinate Kai Altmann, the Chief Engineer. Fail.
Objective #2: Stay alive until the end. Fail.
The traitor has failed!

I wish I could provide more details about what the fuck happened, but I don't have them. No idea. Literally none. All I have is the evidence. There's https://github.com/tgstation/-tg-station/issues/1796 but that might be about traitor borgs. And this seems to be about malf: https://github.com/tgstation/-tg-station/issues/2915

### LINKED ISSUE #13337: Borgs sometimes not recieving law 0.
Problem Description: Cyborgs slaved to traitor AI's sometimes do not recieve the modified lawset, resulting in asimov borgs being ordered to do traitor things (Causing a lot of confusion)

What did you expect to happen: Normally borgs slaved to traitor ai's will have a law 0.

What happened instead: No law 0 was found (Though I think some of the description around the laws still say you have to help with his objectives?)

Why is this bad/What are the consequences: Because the immediate reaction by asimov borgs when ai tells them to kill X person is often to shout in common "AI IS MALF TRYING TO KILL X!"

Steps to reproduce the problem: I never did anything special except spawn as roundstart cyborg and the ai was a traitor.

Server: Sybil

Revision: 2015-11-30
0142846370eb1abf3bdb20b2ae51c6fe69cc1d9f 

Possibly related stuff (which gamemode was it? What were you doing at the time? Was anything else out of the ordinary happening?): As far as I could tell it was traitor, or maybe traitorlings.

### LINKED ISSUE #14241: Traitor objective / law 0 from traitor AI not given to roundstart cyborg
**Server:** Sybil

**Revision:** Didn't grab the exact number; it's whatever revision was active around 2015-12-31 20:30 UTC+01:00.

**Problem Description:** I was the roundstart cyborg. Initial law display listed Asimov, with no special messages or traitor announcements. Beep over to Robotics to get a replacement cell. Roboticist pops me open, and right around this point I notice a 'Hacked' icon in the top left of the screen. Assuming the Roboticist just emagged me, I check the chat log for the usual emag messages, but there are none. 

I click Show Laws, which lists standard Asimov again, but also says something about the AI not knowing about my law 0, and to avoid suspicion. It also says obeying is optional, and that the laws are synched to the AI. There is no zeroth law listed, and my notes don't show any objectives.

At this point, the Roboticist notices the busted interface and decons me. I adminhelp asking what the hell, STP looks into it and mentions that both the AI and the newly created cyborg had a law 0. He couldn't check my laws anymore since I was deconned.

**What did you expect to happen:** A zeroth law, objectives, or any indication whatsoever that I was apparantly a traitor and/or slaved to a traitorous AI.

**What happened instead:** Standard Asimov, no special messages at all; only the 'Show Laws' verb giving a weird line about a non-existent law 0.

**Why is this bad/What are the consequences:** Same as the older incarnations of this issue. Should be clear enough from the story.

**Steps to reproduce the problem:** I'm guessing it's not a 100% thing, or it might have been encountered sooner, so this is pretty bare-bones.

1. Get roundstart cyborg
2. Get a traitor AI

**Possibly related stuff:** The Show-Laws message was implying that I was a lone traitorous cyborg, and that the AI was kosher. STP mentioned that the AI definately had a zeroth law, so this seems to be contradictory.

### PR DESCRIPTION
**Pull request by Supermichael777**
re-based and rewritten.
Fixes #14241
Fixes #13337
Fixes #9679
Fixes #2915 (sorta the same problem but the specific problem code is gone)
Closes: #14272 (this is a much beter way)

### CODE REVIEW

**Comment by Supermichael777** [comment]
> 13 commits

**Comment by WJohn** [comment]
Quad-kill. Brutal.

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 1]
```diff
@@ -1169,4 +1169,7 @@
 				src << "<span class='warning'>SYSTEM ERROR: Module 2 OFFLINE.</span>"
 			if(health < -maxHealth*0.5)
 				if(uneq_module(module_state_1))
-					src << "<span class='warning'>CRITICAL ERROR: All modules OFFLINE.</span>"
\ No newline at end of file
+					src << "<span class='warning'>CRITICAL ERROR: All modules OFFLINE.</span>"
+
+/mob/living/silicon/robot/proc/synchborgs()
+	return
```
Comment (tkdrg) [inline code/modules/mob/living/silicon/robot/robot.dm]: Why did you do this?

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 2]
```diff
@@ -848,4 +848,13 @@ var/list/ai_list = list()
...
+	show_message(rendered, 2)
+
+/mob/living/silicon/ai/proc/synchborgs()
+	for(var/mob/living/silicon/robot/I in connected_robots)
+		if(I.lawupdate)
+			I.lawsync()
+			I << "From now on, these are your laws:"
+			I.show_laws()
+			I.law_change_counter++
+	return
```
Comment (tkdrg) [inline code/modules/mob/living/silicon/ai/ai.dm]: Unneeded but not that big of a deal

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 1 above)
Comment (Supermichael777) [inline code/modules/mob/living/silicon/robot/robot.dm]: I don't going to take it out

**Comment by Supermichael777** [comment]
this probably won't be finalized for a half a day or so

**Comment by Supermichael777** [comment]
in its current state its done, but I may remove that bit of unnessicary code. wish I could squash this but I am not allowed to install git on the computer I'm using (long story)

**Comment by Supermichael777** [comment]
Right, its done unless something needs fixing.

**Comment by KorPhaeron** [comment]
@Aranclanos merged my fix for this, this is redundant now and would end up displaying laws twice for the cyborgs

**Comment by Aranclanos** [comment]
try to make a single commit before you make a PR instead of making 1 commit each time that you edit the file, that way you won't have a 15 commit PR. Also name them properly (basically the PR text)
all issues fixed already so whatevs

**Pull request closed without merging**